### PR TITLE
feat: tldr to article pages

### PR DESCRIPTION
**Pull request by ayadotdev**
Added TLDR to the article page. When the TLDR content is more than 300 characters "Show more" link is displayed. When the content is expanded "Show less" link is displayed. When the content is less than 300 characters, "Show more" link is not displayed.

Screenshot 2022-01-18 at 15 54 43

Screenshot 2022-01-18 at 15 55 22

Screenshot 2022-01-18 at 15 59 50

DD-279 #done

### CODE REVIEW

File: packages/shared/src/components/buttons/ClickableText.tsx [hunk 1]
```diff
@@ -21,6 +23,8 @@ function ClickableTextComponent<Tag extends AvailableTags>(
     children,
     tag: Tag = 'button',
     defaultTypo = true,
+    flexRowApplied = true,
+    underlined = true,
```
Comment (rebelchris) [inline packages/shared/src/components/buttons/ClickableText.tsx:27]: I think we are actually better of keeping this outside the component.
This specific spec guideline uses it, but the overal one doesn't.

You can set classNames for this specific use-case like so:

```
<ClickableText className="flex flex-row">
```

I think you might have noticed it's not inline because the default tag is a button.
You can change it to a link like so:

```
<ClickableText tag="a" />
```

File: packages/shared/src/components/cards/PostTLDRSummary.tsx [hunk 2]
```diff
@@ -0,0 +1,26 @@
+import React, { ReactElement } from 'react';
+import { Post } from '../../graphql/posts';
+import { TLDRSummary } from '../utilities';
+import ShowMoreContent from './ShowMoreContent';
+
+interface PostTLDRSummaryProps {
+  post: Post;
+}
+
+export default function PostTLDRSummary({
```
Comment (rebelchris) [inline packages/shared/src/components/cards/PostTLDRSummary.tsx:10]: I think PostSummary might be a better name here.

File: packages/shared/src/components/cards/PostTLDRSummary.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  post,
+}: PostTLDRSummaryProps): ReactElement {
+  return (
+    <TLDRSummary>
```
Comment (rebelchris) [inline packages/shared/src/components/cards/PostTLDRSummary.tsx:14]: This might also be more inline if it's SummaryContainer or something like that?
What do you think?

File: packages/shared/src/components/cards/ShowMoreContent.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        {isTextExpanded ? content : getSlicedContent()}
+        &nbsp;
+        <ClickableText
+          className="text-theme-label-link"
+          flexRowApplied={false}
+          underlined={false}
+          onClick={toggleTextExpanded}
+        >
+          {displayShowMoreLink()}
+        </ClickableText>
```
Comment (rebelchris) [inline packages/shared/src/components/cards/ShowMoreContent.tsx:47]: It would be better to completely not render this if it's not applicable.

File: packages/shared/src/components/cards/ShowMoreContent.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    }
+    return '';
+  };
+
+  const getSlicedContent = () => {
+    return content?.slice(0, charactersLimit);
+  };
+  return (
+    <div className={className}>
+      <p className="typo-callout">
```
Comment (rebelchris) [inline packages/shared/src/components/cards/ShowMoreContent.tsx:36]: I think this should be `typo-headline` looking at the spec.

Screenshot 2022-01-19 at 09 52 23

File: packages/shared/src/components/fields/Dropdown.tsx [hunk 6]
```diff
@@ -30,7 +30,7 @@ export interface DropdownProps {
 
 const getButtonSizeClass = (buttonSize: string): string => {
   if (buttonSize === 'select') {
-    return 'h-9 rounded-10 text-theme-label-primary typo-callout';
+    return 'h-9 rounded-10 text-theme-label-primary typo-body';
```
Comment (rebelchris) [inline packages/shared/src/components/fields/Dropdown.tsx:33]: Why did you change this one?

File: packages/shared/src/components/fields/TextField.module.css [hunk 7]
```diff
@@ -10,7 +10,6 @@
 
   &:global(.invalid) {
     box-shadow: inset 0.125rem 0 0 0 var(--theme-status-error);
-    border-color: var(--theme-status-error);
```
Comment (rebelchris) [inline packages/shared/src/components/fields/TextField.module.css:13]: Also not sure why this one is removed here?

File: packages/shared/src/components/fields/TextField.tsx [hunk 8]
```diff
@@ -120,7 +120,7 @@ export function TextField({
     >
       {compact && (
         <label
-          className="px-2 mb-1 font-bold text-theme-label-primary"
```
Comment (rebelchris) [inline packages/shared/src/components/fields/TextField.tsx:123]: I think you might have mixed two Pr's here

File: packages/shared/src/components/utilities.tsx [hunk 9]
```diff
@@ -89,3 +89,11 @@ export const customFeedIcon = 'text-2xl text-theme-label-tertiary mr-2';
...
 
 export const SummaryArrow = classed(ArrowIcon, 'icon arrow ml-auto text-xl');
+
+export const TLDRSummaryBorder = 'border-l border-theme-status-cabbage pl-4';
+
+export const TLDRSummary = classed(
+  'div',
+  'text-theme-label-secondary multi-truncate',
+  TLDRSummaryBorder,
+);
```
Comment (rebelchris) [inline packages/shared/src/components/utilities.tsx:99]: I would combine these two, don't think we need them seperate so:

```js
export const TLDRSummary = classed(
  'div',
  'text-theme-label-secondary multi-truncate border-l border-theme-status-cabbage pl-4',
);
```

Comment (rebelchris) [reply packages/shared/src/components/utilities.tsx:99]: Also this should have the top and bottom margin as Tsahi mentioned.

**Review by rebelchris — CHANGES_REQUESTED** [review]
Really nice work Aya!
Have some remarks, but love this setup. 🥳

File: packages/shared/src/components/fields/common.tsx [hunk 10]
```diff
@@ -3,7 +3,7 @@ import styles from './fields.module.css';
 
 export const FieldInput = classed(
   'input',
-  'min-w-0 text-theme-label-primary bg-transparent typo-callout caret-theme-label-link focus:outline-none',
+  'min-w-0 text-theme-label-primary bg-transparent typo-body caret-theme-label-link focus:outline-none',
```
Comment (sshanzel) [inline packages/shared/src/components/fields/common.tsx:6]: Tagging changes in this file as well.

File: packages/shared/src/components/profile/ProfileForm.tsx [hunk 11]
```diff
@@ -280,7 +280,7 @@ export default function ProfileForm({
         }
       />
       <div className="flex flex-col items-stretch self-stretch mb-3">
-        <p className="px-2 mb-1 font-bold text-theme-label-primary">
+        <p className="px-2 mb-1 font-bold text-theme-label-primary typo-caption1">
```
Comment (sshanzel) [inline packages/shared/src/components/profile/ProfileForm.tsx:283]: Tagging changes in this file as well.

File: packages/webapp/pages/posts/[id].tsx [hunk 12]
```diff
@@ -243,23 +244,30 @@ const PostPage = ({ id, postData }: Props): ReactElement => {
...
           <PostHeader post={postById.post} />
-
           <a {...postLinkProps} className="cursor-pointer">
-            <h1 className="my-2 font-bold typo-title2">
+            <h1 className="my-2 font-bold typo-large-title">
               {postById?.post.title}
             </h1>
           </a>
-
+          <PostTLDRSummary post={postById.post} />
```
Comment (sshanzel) [inline packages/webapp/pages/posts/[id].tsx:270]: I think we should dynamically render this only if the `post.summary` is actually available - this would display the TLDR even summary should not be available.

File: packages/shared/src/components/buttons/ClickableText.tsx [hunk 13]
```diff
@@ -8,6 +8,8 @@ export interface BaseClickableTextProps {
   pressed?: boolean;
   tag?: React.ElementType & AvailableTags;
   defaultTypo?: boolean;
+  flexRowApplied?: boolean;
+  underlined?: boolean;
```
Comment (sshanzel) [inline packages/shared/src/components/buttons/ClickableText.tsx:12]: These properties will violate the guidelines we have for `ClickableText`. I think this shouldn't be necessary.

**Review by sshanzel — CHANGES_REQUESTED** [review]
Wow, super amazing you took this task. Very exciting to see this get merged.

I had quite a few comments but you are on the right track! It was mostly the changes that shouldn't be made since those were not related to the task at hand.

👏

File: packages/shared/src/components/buttons/ClickableText.tsx (hunk 1 above)
Comment (rebelchris) [reply packages/shared/src/components/buttons/ClickableText.tsx:27]: Oh and as extra: The underline already is applied to the default ClickableText element on hover.

File: packages/shared/src/components/cards/PostTLDRSummary.tsx (hunk 3 above)
Comment (ayadotdev) [reply packages/shared/src/components/cards/PostTLDRSummary.tsx:14]: It makes sense 👍

File: packages/shared/src/components/cards/ShowMoreContent.tsx (hunk 5 above)
Comment (ayadotdev) [reply packages/shared/src/components/cards/ShowMoreContent.tsx:36]: I applied `typo-headline` before but with that all text becomes bold:

Screenshot 2022-01-19 at 13 49 23

File: packages/shared/src/components/cards/PostTLDRSummary.tsx [hunk 14]
```diff
@@ -0,0 +1,26 @@
+import React, { ReactElement } from 'react';
+import { Post } from '../../graphql/posts';
+import { TLDRSummary } from '../utilities';
+import ShowMoreContent from './ShowMoreContent';
+
+interface PostTLDRSummaryProps {
+  post: Post;
```
[the remark this answers is not in the document]
Comment (ayadotdev) [reply packages/shared/src/components/cards/PostTLDRSummary.tsx:7]: Nice one, Lee!

File: packages/shared/src/components/cards/ShowMoreContent.tsx (hunk 5 above)
Comment (rebelchris) [reply packages/shared/src/components/cards/ShowMoreContent.tsx:36]: Hmm, you might have to double check with Tsahi.
I think it should be 17px, but maybe not bold.

Comment (ayadotdev) [reply packages/shared/src/components/cards/ShowMoreContent.tsx:36]: There is a bold font-weight in typo-headline:
```
 '.typo-headline': {
      'font-size': '1.0625rem',
      'line-height': '1.5rem',
      'font-weight': 'bold',
    }
```

Comment (ayadotdev) [reply packages/shared/src/components/cards/ShowMoreContent.tsx:36]: Clarified with @tsahimatsliah, it should be typo-body.

File: packages/shared/src/components/fields/TextField.tsx [hunk 15]
```diff
@@ -137,6 +137,7 @@ export function TextField({
           },
           styles.field,
         )}
+        style={{ ...style, borderRadius: 'rounded-10' }}
```
Comment (rebelchris) [inline packages/shared/src/components/fields/TextField.tsx:140]: This think this is not right?
Not sure this is even needed here

File: packages/shared/src/components/utilities.tsx [hunk 16]
```diff
@@ -89,3 +89,9 @@ export const customFeedIcon = 'text-2xl text-theme-label-tertiary mr-2';
 export const Summary = classed('summary', 'cursor-pointer focus-outline');
 
 export const SummaryArrow = classed(ArrowIcon, 'icon arrow ml-auto text-xl');
+
+export const SummaryContainer = classed(
+  'div',
+  'text-theme-label-secondary multi-truncate my-6',
+  'border-l border-theme-status-cabbage pl-4',
```
Comment (rebelchris) [inline packages/shared/src/components/utilities.tsx:96]: I would combine these in one line

File: packages/shared/src/components/cards/ShowMoreContent.tsx [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+
+  const toggleTextExpanded = () => {
+    setIsTextExpanded(!isTextExpanded);
+  };
+
+  const displayShowMoreLink = () => {
+    if (
+      content &&
+      content?.length > charactersLimit &&
+      content?.length - charactersLimit > threshold
```
Comment (rebelchris) [inline packages/shared/src/components/cards/ShowMoreContent.tsx:30]: Think we can do this in one go:

```js
content?.length > charactersLimit + threshold
```

**Review by rebelchris — CHANGES_REQUESTED** [review]
Really nice work!
Some small things, almost there 🥳

File: packages/shared/src/components/fields/TextField.tsx (hunk 15 above)
Comment (ayadotdev) [reply packages/shared/src/components/fields/TextField.tsx:140]: Weird, I removed it yesterday.

**Review by rebelchris — APPROVED** [review]
Super nice work you did here! 🎉
Love the tests

File: packages/shared/src/components/cards/ShowMoreContent.tsx [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+  const toggleTextExpanded = () => {
+    setIsTextExpanded(!isTextExpanded);
+  };
+
+  const displayShowMoreLink = () => {
+    if (content && content?.length > charactersLimit + threshold) {
+      return true;
+    }
+    return false;
+  };
```
Comment (idoshamun) [inline packages/shared/src/components/cards/ShowMoreContent.tsx:31]: It looks cleaner in my opinion

```suggestion
  const displayShowMoreLink = () => content && content?.length > charactersLimit + threshold;
```

File: packages/shared/src/components/cards/ShowMoreContent.tsx [hunk 19]
```diff
@@ -0,0 +1,53 @@
...
+  threshold = 50,
+  contentPrefix,
+  className,
+}: ShowMoreContentProps): ReactElement {
+  const [isTextExpanded, setIsTextExpanded] = useState(false);
+  const linkName = isTextExpanded ? 'Show less' : 'Show more';
+
+  const toggleTextExpanded = () => {
+    setIsTextExpanded(!isTextExpanded);
+  };
```
Comment (idoshamun) [inline packages/shared/src/components/cards/ShowMoreContent.tsx:24]:
```suggestion
  const toggleTextExpanded = () => setIsTextExpanded(!isTextExpanded);
```

File: packages/shared/src/components/cards/ShowMoreContent.tsx [hunk 20]
```diff
@@ -0,0 +1,53 @@
...
+  const displayShowMoreLink = () => {
+    if (content && content?.length > charactersLimit + threshold) {
+      return true;
+    }
+    return false;
+  };
+
+  const getContent = () => {
+    return isTextExpanded ? content : content?.slice(0, charactersLimit);
+  };
```
Comment (idoshamun) [inline packages/shared/src/components/cards/ShowMoreContent.tsx:35]:
```suggestion
  const getContent = () => (isTextExpanded ? content : content?.slice(0, charactersLimit));
```

File: packages/shared/src/components/cards/ShowMoreContent.tsx [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+  return (
+    <div className={className}>
+      <p className="typo-body">
+        {contentPrefix}
+        {getContent()}{' '}
+        {displayShowMoreLink() && (
+          <ClickableText
+            className="inline-flex text-theme-label-link"
+            onClick={toggleTextExpanded}
+            tag="a"
```
Comment (idoshamun) [inline packages/shared/src/components/cards/ShowMoreContent.tsx:45]: Why is that a `a` element? I think the `tag` property can be dropped

File: packages/shared/src/graphql/posts.ts [hunk 22]
```diff
@@ -103,6 +104,7 @@ export const POST_BY_ID_QUERY = gql`
         username
       }
       description
+      summary
```
Comment (idoshamun) [inline packages/shared/src/graphql/posts.ts:107]: We need to add this to `POST_BY_ID_STATIC_FIELDS_QUERY` so it will be cached for the next visits.
In the article page itself you can find this function `getStaticProps` that fetches the necessary data for the page before it arrives to the client and then we revalidate on the client side as well. I highly suggest reading about nextjs's incremental static regeneration for better understanding what's going on under the hood.

@rebelchris @sshanzel  this can be another valid point for future reviews in pages that use ISR or SSR.

**Review by idoshamun — CHANGES_REQUESTED** [review]
Some cosmetics suggestion and an important change.

Overall this looks solid and this is also quite advanced, it's so great to see your progress from day to day. 💪

File: packages/shared/src/components/cards/ShowMoreContent.tsx (hunk 21 above)
Comment (rebelchris) [reply packages/shared/src/components/cards/ShowMoreContent.tsx:45]: I'm pretty sure this should render as a <a> not a <button> right?
Since it's an inline anchor element.

The button would also push it to it's own line.

Comment (idoshamun) [reply packages/shared/src/components/cards/ShowMoreContent.tsx:45]: Hmm interesting. The second argument could be easily fixed but the first one is interesting.
I think anchors should only be links and this one is actually a button. So in terms of accessibility and semantics, I do think it's a button but you made me somewhat question it. Could you check it, please?

Comment (rebelchris) [reply packages/shared/src/components/cards/ShowMoreContent.tsx:45]: Screenshot 2022-01-20 at 13 42 28

To me this is really a <a href=""> not a <button>

But let me know if you think otherwise

Comment (rebelchris) [reply packages/shared/src/components/cards/ShowMoreContent.tsx:45]: But according to A11Y rules you are 100% right and it should be a button (my bad)
Screenshot 2022-01-20 at 13 44 34

Comment (rebelchris) [reply packages/shared/src/components/cards/ShowMoreContent.tsx:45]: Reading a bit more, there seems to be quite the split thinking about these and some field experts even say it should not even exist in this way.

https://blog.prototypr.io/youre-not-still-using-read-more-are-you-16466ae326e7

Comment (idoshamun) [reply packages/shared/src/components/cards/ShowMoreContent.tsx]: Let's go with a button then :)

**Review by sshanzel — APPROVED** [review]
LGTM! 🚀

File: packages/shared/src/graphql/posts.ts (hunk 22 above)
Comment (rebelchris) [reply packages/shared/src/graphql/posts.ts:107]: On this topic, I get the difference, but why can't they use the same GraphQL endpoint?

Comment (idoshamun) [reply packages/shared/src/graphql/posts.ts:107]: Because some of the properties depend on the logged-in user, this will not be available and might even throw an exception. We can try that of course.

**Review by idoshamun — APPROVED** [review]
Let's ship!
